fury/stream/widget.py

```diff
@@ -0,0 +1,254 @@
+import subprocess
+import numpy as np
+import sys
+if sys.version_info.minor >= 8:
+    PY_VERSION_8 = True
+else:
+    PY_VERSION_8 = False
+try:
+    from IPython.display import IFrame
+    from IPython.core.display import display
+    IPYTHON_AVAILABLE = True
+except ImportError:
+    IPYTHON_AVAILABLE = False
+
+import time
+import socket
+import errno
+
+from fury.stream.client import FuryStreamClient, FuryStreamInteraction
+
+
+def check_port_is_available(host, port):
+    """Check if a given port it's available
+
+    Parameters
+    ----------
+    host : str
+    port : int
+
+    Returns
+    -------
+    available : bool
+
+    """
+    available = True
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.bind((host, port))
+    except socket.error as e:
+        if e.errno == errno.EADDRINUSE:
+            available = False
+    s.close()
+    return available
+
+
+class Widget:
+    def __init__(
+            self, showm, ms_stream=33,
+            ms_interaction=33,
+            host='localhost', port=None,
+            encoding='mjpeg', ms_jpeg=33,
+            queue_size=20):
+        """Thi Obj it's able execute the fury streaming system
+        using the SharedMemory from python multiprocessing.
+
+        Parameters
+        ----------
+        showm : ShowmManager
+        ms_stream : float, optional
+            time in mileseconds between each frame buffer update.
+        ms_interaction : float, optional
+            time in mileseconds between each user interaction update.
+        host : str, optional
+        port : int, optional
+        encoding : str, optional
+            If should use MJPEG streaming or WebRTC.
+        ms_jpeg : float, optional
+            This it's used  only if the MJPEG will be used. The
+            ms_jpeg represents the amount of miliseconds between to
+            consecutive calls of the jpeg enconding.
+        queue_size : int, optional
+            maximum number of user interactions to be stored
+
+        """
+        if not PY_VERSION_8:
+            raise ImportError('Python 3.8 or greater is required to use the\
+                widget class')
+        self.showm = showm
+        self.window_size = self.showm.size
+        max_window_size = (
+            int(self.window_size[0]*(1+0.1)),
+            int(self.window_size[1]*(1+0.1))
+        )
+        self.max_window_size = max_window_size
+        self.ms_stream = ms_stream
+        self.ms_interaction = ms_interaction
+        self.ms_jpeg = ms_jpeg
+        self._host = host
+        if port is None:
+            port = np.random.randint(7000, 8888)
```

Comment:
Is this safe to do?

Reply:
Yes, I think so. @filipinascimento  do you have any considerations?

Reply:
This do not check for existing open ports, it is safe, but can fail if the selected port is already open.

Reply:
We can check this. But this operation takes some time.